### PR TITLE
238 refactore api

### PR DESCRIPTION
**Pull request by smebrahimi**
This issue #238 
Changes:

1. api/v0/volkswagen/articles/dailyplan is removed use api/v0/volkswagen/jumpstart/dailyplan instead

2. For JumpStart preview use api/v0/volkswagen/jumpstart/{date}/preview?articleIds= instead of
api/v0/volkswagen/jumpstart/{jumpStartId}/preview?articleIds=

3. To edit or create a JumpStart use POST api/v0/volkswagen/jumpstart

4. api/v0/volkswagen/jumpstart/{jumpStartId}/availablearticles is removed

--------------
fixing position of article in rearrange (It was better to remove the availablearticles so that this part would change sooner, It still needs to get better)

### CODE REVIEW

File: src/components/Email/SelectEmailDate.vue [hunk 1]
```diff
@@ -74,54 +74,45 @@ export default {
...
-          article = selectedJumpstart.articles[i];
-          if (article) this.update_selected_articles(article);
-          else break;
-        }
-        this.debounced_preview();
+    prefillArticles() {
+      let selectedJumpstart;
+      const selectedDate = this.get_email_date;
+      selectedJumpstart = this.get_daily_plan.find(item => {
+        return formatISODate(item.dateTime) === formatISODate(selectedDate);
```
Comment (mort3za) [inline src/components/Email/SelectEmailDate.vue:109]: Using formatISODate on both sides of this equation does not make sense. I suggest using `isSameDay` from date-fns.

File: src/components/JumpStartListItem.vue [hunk 2]
```diff
@@ -3,7 +3,7 @@
     <h2
       class="flex tg-body-emphasis-mobile lg:tg-body-emphasis-desktop whitespace-no-wrap truncate text-left em-high py-4 pl-2"
     >
-      {{ formatDate(model.date, 'EEEE, d MMM,yyyy') }}
+      {{ formatDate(model.dateTime, 'EEEE, d MMM,yyyy') }}
```
Comment (mort3za) [inline src/components/JumpStartListItem.vue:6]: You might need to add a space character after `MMM,`

File: src/store/modules/email.js [hunk 3]
```diff
@@ -10,88 +10,103 @@ export default {
...
+        }
+      }
+    },
     debounced_preview: debounce(
       function({ dispatch }) {
         dispatch('update_preview_link');
       },
-      3000,
-      { maxWait: 3000 }
+      1500,
```
Comment (mort3za) [inline src/store/modules/email.js:108]: We need debounce to be 3s, getting a preview is an expensive job in server.

File: src/views/BlueDeltaRearrange.vue [hunk 4]
```diff
@@ -66,32 +66,39 @@ export default {
...
-      this.update_available_articles(article);
+      this.update_selected_active_articles(article);
       this.update_preview_link('');
       this.debounced_preview();
     },
     isActive(article) {
-      return this.get_selected_articles.indexOf(article);
+      let i, n, x;
+      x = -1;
+      if (article.isActive) {
```
Comment (mort3za) [inline src/views/BlueDeltaRearrange.vue:93]: Please add a comment for complex sections and explain the functionality, or you could move this part to another function.

Comment (atharva3010) [reply src/views/BlueDeltaRearrange.vue:93]: Yeah, on first look I am not really sure what is happening here either.

**Comment by paulchrisluke** [comment]
I have tested this issue on this deploy preview: https://deploy-preview-258--eloquent-thompson-534232.netlify.app

- In the /blue-delta/edit/ page, recipients and scheduled times are not saved after selection/pressing save button 

- Blue Delta preview date is 1 day before scheduled Blue Delta date 
unknown (2)

- Date/Time validation for event category article is shown even though there's something there
image

- Unable to open /blue-delta/edit/  for blue delta emails scheduled for June

**Comment by smebrahimi** [comment]
> I have tested this issue on this deploy preview: https://deploy-preview-258--eloquent-thompson-534232.netlify.app
> 
> * In the /blue-delta/edit/ page, recipients and scheduled times are not saved after selection/pressing save button
> * Blue Delta preview date is 1 day before scheduled Blue Delta date
>   unknown (2)
> * Date/Time validation error for creating new event category article is shown even though there's something there
>   image
> * Unable to open /blue-delta/edit/  for blue delta emails scheduled for June

Some issue above should be put another task

File: src/store/modules/email.js [hunk 5]
```diff
@@ -62,23 +51,22 @@ export default {
...
         state.preview_link = payload;
         return false;
       }
-      const id = state.selected_plan.jumpStartId || state.selected_jumpstart.id;
-      const base = `${BASE_API}/api/v0/volkswagen/jumpstart/${id}/preview`;
+      const base = `${BASE_API}/api/v0/volkswagen/jumpstart/${state.selected_plan.dateTime}/preview`;
       const url = new URL(base);
       state.selected_articles.forEach(article => {
-        url.searchParams.append('articleIds', article.id);
+        article.isActive === undefined || article.isActive ? url.searchParams.append('articleIds', article.id) : false;
```
Comment (atharva3010) [inline src/store/modules/email.js:62]: is there a need for `isActive` now?

Comment (smebrahimi) [reply src/store/modules/email.js]: @atharva3010 Fixed

File: src/components/Email/SelectEmailDate.vue [hunk 6]
```diff
@@ -74,54 +75,44 @@ export default {
...
       });
+      if (selectedPlan) {
+        this.update_selected_plan(selectedPlan);
+        this.update_available_articles(selectedPlan.articles);
+        this.update_selected_articles();
+        for (let i = 0; i < 5; i++) {
+          if (selectedPlan.articles[i]) {
+            this.update_selected_articles(selectedPlan.articles[i]);
+          } else break;
+        }
```
Comment (abdullahcalisir12) [inline src/components/Email/SelectEmailDate.vue:113]: I'm not sure what this 5  means and if there are more than 5 articles in selectedPlan object this code will work partly.
I'm sorry I don't know about the work flow but this seemed strange to me.

File: src/views/JumpStartLists.vue [hunk 7]
```diff
@@ -29,22 +29,24 @@ export default {
...
     selectPlan(plan) {
-      this.update_selected_articles();
       this.update_selected_plan(plan);
-      plan.articles.forEach(plan => {
-        this.update_selected_articles(plan);
-      });
-      this.update_email_date(plan.date);
+      this.update_available_articles(plan.articles);
+      this.update_selected_articles();
+      for (let i = 0; i < 5; i++) {
```
Comment (abdullahcalisir12) [inline src/views/JumpStartLists.vue:44]: same here

File: src/components/TodayEmail.vue [hunk 8]
```diff
@@ -29,12 +29,12 @@ export default {
...
+    this.fetch_daily_plan()
       .then(() => {
         let article, i;
-        this.update_selected_jumpstart(this.get_jumpstarts[0]);
+        this.update_selected_plan(this.get_daily_plan[0]);
         for (i = 0; i < 5; i++) {
-          article = this.get_selected_jumpstart.articles[i];
+          article = this.get_selected_plan.articles[i];
           if (article) this.update_selected_articles(article);
           else break;
```
Comment (abdullahcalisir12) [inline src/components/TodayEmail.vue:39]: same here too

**Comment by atharva3010** [comment]
PR looks good @smebrahimi! Great work!
@paulchrisluke @mort3za @bamboochow do you have any more feedback? If no, then I'll merge.

**Comment by paulchrisluke** [comment]
Good to merge on our side

File: src/components/Email/SelectEmailDate.vue (hunk 6 above)
Comment (atharva3010) [reply src/components/Email/SelectEmailDate.vue:113]: We only allow 5 articles to be selected for a day’s blue delta. The template can only support 5 at the moment, once we have support for multi page templates, we’ll remove this.

File: src/components/TodayEmail.vue (hunk 8 above)
Comment (atharva3010) [reply src/components/TodayEmail.vue:39]: The 5 articles which are first in order among all the available articles are pushed to the selected articles array

File: src/views/JumpStartLists.vue (hunk 7 above)
Comment (atharva3010) [reply src/views/JumpStartLists.vue:44]: It’s similar to what’s explained above

**Comment by atharva3010** [comment]
Great then! Merging PR

File: src/components/Email/SelectEmailDate.vue (hunk 6 above)
Comment (abdullahcalisir12) [reply src/components/Email/SelectEmailDate.vue:113]: it would be easier to use forEach or instead of 5 we could use selected_articles length to iterate

Comment (abdullahcalisir12) [reply src/components/Email/SelectEmailDate.vue:113]: or MAX_SELECTED_ARTICLE_NUMBER to constant and using it

Comment (atharva3010) [reply src/components/Email/SelectEmailDate.vue:113]: forEach iterates through all the elements of the array. We need to select only the first five elements, so iterating through rest of the elements is unnecessary.

However, creating a shallow copy of the articles and then splicing all the items after index 4 can be a viable option.

The requirement of 5 articles will be removed soon though and we can remove this altogether.

**Pull request merged**